### PR TITLE
feat: Add "Comet Fuzz" fuzz-testing utility

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

Comet Fuzz is a standalone project for generating random data and queries and executing queries against Spark 
with Comet disabled and enabled and checking for incompatibilities.

Although it is a simple tool it has already been useful in finding many bugs, including:

- #479 
- #480 
- #481 
- #482 
- #483 
- #484 
- #485 
- #486 

Comet Fuzz is inspired by the [SparkFuzz](https://ir.cwi.nl/pub/30222) paper from Databricks and CWI.

## What changes are included in this PR?

New standalone `comet-fuzz` project.

## How are these changes tested?

Manually tested.

### CODE REVIEW

File: dev/scalastyle-config.xml [hunk 1]
```diff
@@ -242,7 +242,7 @@ This file is divided into 3 sections:
       <parameter name="groups">java,scala,org,apache,3rdParty,comet</parameter>
       <parameter name="group.java">javax?\..*</parameter>
       <parameter name="group.scala">scala\..*</parameter>
-      <parameter name="group.org">org\.(?!apache\.comet).*</parameter>
```
Comment (andygrove) [inline dev/scalastyle-config.xml:245]: We should exclude all apache packages from the `org` group, not just comet. We have a separate rule below for excluding comet from the apache group.

Without this change it was not possible to satisfy both scalastyle and scalafix at the same time as they wanted different import orders.

File: fuzz-testing/pom.xml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-sql_${scala.binary.version}</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.rogach</groupId>
+            <artifactId>scallop_${scala.binary.version}</artifactId>
+            <version>5.1.0</version>
+        </dependency>
```
Comment (kazuyukitanimura) [inline fuzz-testing/pom.xml:58]: nit: (optional) we use `<dependencyManagement>` for version

File: fuzz-testing/src/main/scala/org/apache/comet/fuzz/DataGen.scala [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  def generateRandomParquetFile(
+      r: Random,
+      spark: SparkSession,
+      filename: String,
+      numRows: Int,
+      numColumns: Int): Unit = {
+
+    // TODO add examples of all supported types, including complex types
+    val dataTypes = Seq(
+      (DataTypes.ByteType, 0.2),
```
Comment (kazuyukitanimura) [inline fuzz-testing/src/main/scala/org/apache/comet/fuzz/DataGen.scala:51]: What about adding boolean and binary types?

**Comment by andygrove** [comment]
Thanks for the review @kazuyukitanimura. I have addressed your comments and also added support for decimal type. I  removed the TODO comments and added a roadmap section to the README to cover some of the planned future work.

**Comment by andygrove** [comment]
I'm not sure why tests are being cancelled on this PR.

**Comment by viirya** [comment]
> I'm not sure why tests are being cancelled on this PR.

I saw the same on some PRs too. Re-triggering CIs can make them back.

File: fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+  private def generateAggregate(r: Random, spark: SparkSession, numFiles: Int): String = {
+    val tableName = s"test${r.nextInt(numFiles)}"
+    val table = spark.table(tableName)
+
+    val func = Utils.randomChoice(Meta.aggFunc, r)
+    val args = Range(0, func.num_args)
+      .map(_ => Utils.randomChoice(table.columns, r))
+
+    val groupingCols = Range(0, 2).map(_ => Utils.randomChoice(table.columns, r))
```
Comment (viirya) [inline fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala:66]: Do we need to exclude gouping columns from aggregation function arguments?

File: fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    val joinType = Utils.randomWeightedChoice(joinTypes)
+
+    val leftColProjection = leftTable.columns.map(c => s"l.$c").mkString(", ")
+    val rightColProjection = rightTable.columns.map(c => s"r.$c").mkString(", ")
+    "SELECT " +
+      s"$leftColProjection, " +
+      s"$rightColProjection " +
+      s"FROM $leftTableName l " +
+      s"$joinType JOIN $rightTableName r " +
+      s"ON l.$leftCol = r.$rightCol " +
```
Comment (viirya) [inline fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala:113]: `leftCol` and `rightCol` are randomly picked up. Are they always same data types?

Comment (viirya) [reply fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala:113]: Or it doesn't matter and we let Spark automatically cast between them?

File: fuzz-testing/src/main/scala/org/apache/comet/fuzz/Main.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  }
+  addSubcommand(runQueries)
+  verify()
+}
+
+object Main {
+
+  lazy val spark: SparkSession = SparkSession
+    .builder()
+    .master("local[*]")
```
Comment (viirya) [inline fuzz-testing/src/main/scala/org/apache/comet/fuzz/Main.scala:54]: Do we need to hardcode this?

File: fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala (hunk 5 above)
Comment (andygrove) [reply fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala:113]: Exactly. We just pick random columns and let Spark add implicit casts when needed.

File: fuzz-testing/src/main/scala/org/apache/comet/fuzz/Main.scala (hunk 6 above)
Comment (andygrove) [reply fuzz-testing/src/main/scala/org/apache/comet/fuzz/Main.scala]: I have removed this. We specify master during spark-submit so do not need to hardcode it here.

File: fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala (hunk 4 above)
Comment (andygrove) [reply fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala:66]: If it is valid SQL I think we should allow it

**Review by viirya — APPROVED** [review]

**Comment by andygrove** [comment]
Thanks for the review @viirya

**Pull request merged**